### PR TITLE
[ENH] owfile: allow multiple readers with same extension

### PR DESCRIPTION
**Pull request by markotoplak**
##### Issue
Same extension can describe different kinds of data files.

##### Description of changes
Users can explicitly select a reader in the file dialog. If they do not, Orange uses the reader for that extension with the lowest priority.

##### Includes
- [X] Code changes
- [X] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by borondics** [comment]
Could you just simply cycle through the readers and whichever doesn't
produce an error would be the right one?

**Comment by markotoplak** [comment]
@borondics, @astaric had an interesting idea: these changes allow anyone (or an add-on) to register a meta-reader which does just what you propose.

File: Orange/data/io.py [hunk 1]
```diff
@@ -762,6 +763,17 @@ def write_data(cls, write, data):
                    val
                    for var, val in zip(vars, flatten(row))])
 
+    @classmethod
+    def qualified_name(cls):
+        return cls.__module__ + '.' + cls.__name__
+
+
+def file_format_from_qualified_name(format_name):
```
Comment (astaric) [inline Orange/data/io.py]: object_from_qualified_name?

File: Orange/widgets/data/tests/test_owfile.py [hunk 2]
```diff
@@ -209,6 +209,27 @@ def test_check_datetime_disabled(self):
                 vartype_delegate.setEditorData(combo, idx(i))
                 self.assertEqual(combo.count(), counts[i])
 
+    def test_read_format(self):
+        iris = Table("iris")
+
+        def iris_with_no_specific_format(a, b, c, filters, e):
```
Comment (astaric) [inline Orange/widgets/data/tests/test_owfile.py]: open_iris_with_no_specific_format

File: Orange/widgets/data/tests/test_owfile.py [hunk 3]
```diff
@@ -209,6 +209,27 @@ def test_check_datetime_disabled(self):
...
+            return iris.__file__, filters.split(";;")[0]
+
+        with patch("AnyQt.QtWidgets.QFileDialog.getOpenFileName",
+                   iris_with_no_specific_format):
+            self.widget.browse_file()
+
+        self.assertIsNone(self.widget.recent_paths[0].file_format)
+
+        def iris_with_tab(a, b, c, filters, e):
```
Comment (astaric) [inline Orange/widgets/data/tests/test_owfile.py]: open_iris_with_tab

File: Orange/widgets/utils/filedialogs.py [hunk 4]
```diff
@@ -104,14 +104,54 @@ def get_file_name(start_dir, start_filter, file_formats):
...
+    Open file dialog with file formats.
+
+    Args:
+        start_dir (str): initial directory, optionally including the filename
+        start_filter (str): initial filter
+        file_formats (dict {extension: Orange.data.io.FileFormat}): file formats
+        title (str): title of the dialog
+        dialog: a function that creates a QT dialog
+    Returns:
+        (filename, filter, file_format), or `(None, None, None)` on cancel
```
Comment (astaric) [inline Orange/widgets/utils/filedialogs.py]: This does not match
```
return filename, file_format, filter
```

File: Orange/widgets/utils/filedialogs.py [hunk 5]
```diff
@@ -104,14 +104,54 @@ def get_file_name(start_dir, start_filter, file_formats):
...
+    return filename, file_format, filter
+
+
 class RecentPath:
     abspath = ''
     prefix = None   #: Option[str]  # BASEDIR | SAMPLE-DATASETS | ...
     relpath = ''  #: Option[str]  # path relative to `prefix`
     title = ''    #: Option[str]  # title of filename (e.g. from URL)
     sheet = ''    #: Option[str]  # sheet
+    file_format = None  #: Option[str]  # file format as a string
```
Comment (astaric) [inline Orange/widgets/utils/filedialogs.py]: preferred_reader?

**Pull request merged**